### PR TITLE
ARROW-9867: [C++][Dataset] Add FileSystemDataset::filesystem property

### PR DESCRIPTION
**Pull request by bkietz**
In addition, the constructor now ensures that all fragments in a file system dataset belong to the provided filesystem.

### CODE REVIEW

**Review by nealrichardson — COMMENTED** [review]
Thanks for doing this and for adding the R binding. Would you mind adding a test like `expect_is(child$filesystem, "LocalFileSystem")` around L705 of test-dataset.R?

File: cpp/src/arrow/dataset/file_base.cc [hunk 1]
```diff
@@ -82,25 +82,34 @@ Result<ScanTaskIterator> FileFragment::Scan(std::shared_ptr<ScanOptions> options
...
-  return std::shared_ptr<FileSystemDataset>(
-      new FileSystemDataset(std::move(schema), std::move(root_partition),
-                            std::move(format), std::move(fragments)));
+  for (const auto& fragment : fragments) {
+    if ((filesystem == nullptr && fragment->source().filesystem() != nullptr) ||
+        (filesystem != nullptr &&
+         !fragment->source().filesystem()->Equals(*filesystem))) {
+      return Status::Invalid("FileSystemDataset's filesystem differed from a fragment's");
+    }
+  }
```
Comment (jorisvandenbossche) [inline cpp/src/arrow/dataset/file_base.cc:102]: This validation should not be needed in many cases (eg from filesystem discovery, or from ParquetFactory, we know for sure that all fragments are already coming from the same filesystem), so I think we should avoid that when possible.

File: python/pyarrow/_dataset.pyx [hunk 2]
```diff
@@ -467,12 +467,13 @@ cdef class FileSystemDataset(Dataset):
     cdef:
         CFileSystemDataset* filesystem_dataset
 
-    def __init__(self, fragments, Schema schema, FileFormat format,
+    def __init__(self, filesystem, fragments, Schema schema, FileFormat format,
```
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx:470]: Can you add this to the docstring? 
I would also move this keyword after `format`, I think fragments is most logical to come first. 

If we want to have this backwards compatible, we could also make this keyword optional, and if not specified, take it from the first fragment.

File: python/pyarrow/_dataset.pyx [hunk 3]
```diff
@@ -539,7 +549,7 @@ cdef class FileSystemDataset(Dataset):
         ]:
             if not isinstance(arg, class_):
                 raise TypeError(
-                    "Argument '{0}' has incorrect type (expected {1}, "
+                    "Argument '{0}' wtf has incorrect type (expected {1}, "
```
Comment (jorisvandenbossche) [inline python/pyarrow/_dataset.pyx:552]: This was not an intended change .. ? ;-)

**Review by jorisvandenbossche — COMMENTED** [review]
Looks good! My main comment is that I think we should avoid the validation when possible (I don't know how costly this is if you have a huge number of fragments)

Would you like to add a test for the new `filesystem` attribute in Python as well (I can also push that tomorrow)

**Review by nealrichardson — APPROVED** [review]

**Comment by bkietz** [comment]
CI failure is an S3 flake https://github.com/apache/arrow/pull/8069/checks?check_run_id=1042786261#step:8:3006

**Pull request closed without merging**